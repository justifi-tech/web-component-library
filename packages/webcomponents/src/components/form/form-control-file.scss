@import 'reboot';
@import 'forms';
@import 'badge';

:host {
  @include host-base-styles;
  display: block;
}

input[type='file'][disabled] {
  color: var(--jfi-form-control-disabled-color);
  background-color: var(--jfi-form-control-disabled-background-color);
  opacity: 0.5;
  cursor: not-allowed;
}

.input-group-text {
  background-color: var(--jfi-form-control-background-color);
  border-color: var(--jfi-form-control-border-color);
  color: var(--jfi-form-control-color);
}

.file {
  border-top-right-radius: 0 !important;
  border-bottom-right-radius: 0 !important;
}
