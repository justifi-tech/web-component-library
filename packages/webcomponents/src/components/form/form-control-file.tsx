import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  State,
  Element,
} from '@stencil/core';

@Component({
  tag: 'form-control-file',
  styleUrl: 'form-control-file.scss',
  shadow: true,
})
export class FileInput {
  @Element() el: HTMLElement;
  @Prop() label: string;
  @Prop() name: any;
  @Prop() error: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() disabled: boolean;
  @State() input: string;
  @State() file: File;
  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  updateInput(newValue: any) {
    const inputElement = this.el.shadowRoot.querySelector('input');
    if (inputElement) {
      inputElement.value = newValue || '';
    }
  }

  handleFormControlInput(event: any) {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
    this.formControlInput.emit(target.value);
  }

  handleFileChange(event: any) {
    this.file = event.target.files[0];
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <label part="label" class="form-label" htmlFor={this.name}>
          {this.label}
        </label>
        <input
          type="file"
          name={this.name}
          part={`input ${this.error ? 'input-invalid ' : ''}${this.disabled ? ' input-disabled' : ''}`}
          class={this.error ? 'form-control is-invalid' : 'form-control'}
          disabled={this.disabled}
          onChange={(event: any) => this.handleFileChange(event)}
          onInput={(event: any) => this.handleFormControlInput(event)}
          onBlur={() => this.formControlBlur.emit()}
        />
        {this.error && <div class="invalid-feedback">{this.error}</div>}
      </Host>
    );
  }
}
