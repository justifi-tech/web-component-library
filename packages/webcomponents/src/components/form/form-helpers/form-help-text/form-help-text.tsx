import { Component, Prop, h } from '@stencil/core';

@Component({
  tag: 'form-help-text',
})
export class FormHelpText {
  @Prop() helpText: string;
  @Prop() errorText: string;

  private text: string;

  get colorClass() {
    return this.errorText ? 'text-danger' : 'text-muted';
  }

  render() {

    if (!this.helpText && !this.errorText) {
      return null;
    }

    if (this.errorText) {
      this.text = this.errorText;
    } else {
      this.text = this.helpText;
    }

    return (
      <small class={`form-text ${this.colorClass}`} id='formHelpText'>
        {this.text}
      </small>
    );
  }
}
