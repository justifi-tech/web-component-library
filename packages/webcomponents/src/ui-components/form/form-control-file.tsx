import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  State,
  Element,
} from '@stencil/core';
import { EntityDocumentType, FileSelectEvent } from '../../api/Document';
import { FormControlErrorText } from '../../ui-components';
import { FormLabel } from './form-helpers/form-label';

@Component({
  tag: 'form-control-file',
})
export class FileInput {
  @Element() el: HTMLElement;
  fileInput: HTMLInputElement;

  @State() files: File[];
  @State() fileString: string;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() multiple?: boolean;
  @Prop() documentType: EntityDocumentType;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() disabled: boolean;

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;
  @Event() fileSelected: EventEmitter<FileSelectEvent>;

  componentDidLoad() {
    this.fileInput = this.el.querySelector('input');
  }

  handleFormControlInput = (event: any) => {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
    this.formControlInput.emit({ name, value: target.value });
  }

  changeHandler = () => {
    const inputFileList = this.fileInput.files;
    if (inputFileList) {
      this.fileSelected.emit({
        fileList: inputFileList,
        document_type: this.documentType
      });
    }
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <div class="form-group d-flex flex-column">
          <FormLabel
            htmlFor={this.name}
            label={this.label}
            helpText={this.helpText}
          />
          <input
            ref={(el) => this.fileInput = el}
            type="file"
            name={this.name}
            part={`input ${this.errorText ? "input-invalid " : ""}${this.disabled ? "input-disabled" : ""}`}
            class={this.errorText ? "form-control is-invalid" : "form-control"}
            multiple={this.multiple}
            disabled={this.disabled}
            onChange={this.changeHandler}
            onInput={this.handleFormControlInput}
            onBlur={() => this.formControlBlur.emit()}
          />
          <FormControlErrorText errorText={this.errorText} name={this.name} />
        </div>
      </Host>
    );
  }
}
