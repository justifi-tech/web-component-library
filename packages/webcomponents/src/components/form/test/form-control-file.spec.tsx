import { newSpecPage } from '@stencil/core/testing';
import { FileInput } from '../form-control-file';
import { FormHelpText } from '../form-helpers/form-help-text/form-help-text';

describe('form-control-file', () => {

  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [FileInput],
      html: `<form-control-file label="Select a file" name="user ID"></form-control-file>`,
    });

    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.label).toBe('Select a file');
    expect(page.rootInstance.name).toBe('user ID');
  });

  it('renders with all props provided', async () => {
    const page = await newSpecPage({
      components: [FileInput],
      html: `
        <form-control-file
          label="Select a file"
          name="email"
          error="No file selected"
          disabled
        ></form-control-file>
      `,
    });

    const inputElement = page.root.querySelector('input');
    expect(page.rootInstance.label).toBe('Select a file');
    expect(inputElement.disabled).toBeTruthy();
  });

  it('handles user input correctly', async () => {
    const page = await newSpecPage({
      components: [FileInput],
      html: `<form-control-file></form-control-file>`,
    });

    const inputElement = page.root.querySelector('input');
    const testValue = 'Hello, World!';

    inputElement.value = testValue;
    await inputElement.dispatchEvent(new Event('input'));

    expect(inputElement.value).toBe(testValue);
  });

  it('emits formControlInput event on input', async () => {
    const page = await newSpecPage({
      components: [FileInput],
      html: `<form-control-file></form-control-file>`,
    });

    // Set a mock inputHandler to prevent it from being undefined
    page.rootInstance.inputHandler = jest.fn();

    const inputEventSpy = jest.fn();
    page.root.addEventListener('formControlInput', inputEventSpy);

    await page.waitForChanges();

    const inputElement = page.root.querySelector('input');
    inputElement.value = 'Hello, World!';

    inputElement.dispatchEvent(new Event('input', { bubbles: true, composed: true }));

    await page.waitForChanges();

    expect(inputEventSpy).toHaveBeenCalled();
  });

  it('emits formControlBlur event on blur', async () => {
    const page = await newSpecPage({
      components: [FileInput],
      html: `<form-control-file></form-control-file>`,
    });

    const blurEventSpy = jest.fn();
    page.win.addEventListener('formControlBlur', blurEventSpy);

    const inputElement = page.root.querySelector('input');
    inputElement.dispatchEvent(new Event('blur'));

    expect(blurEventSpy).toHaveBeenCalled();
  });

  it('disables input when disabled prop is true', async () => {
    const page = await newSpecPage({
      components: [FileInput],
      html: `<form-control-file disabled></form-control-file>`,
    });

    const inputElement = page.root.querySelector('input');
    expect(inputElement.disabled).toBeTruthy();
  });

  it('shows error and applies error styling when error prop is provided', async () => {
    const page = await newSpecPage({
      components: [FileInput, FormHelpText],
      html: `<form-control-file error-text="This field is required."></form-control-file>`,
    });

    const helpTextComponent = page.root.querySelector('form-help-text');
    expect(helpTextComponent).not.toBeNull();

    const errorText = helpTextComponent.querySelector('.text-danger');
    expect(errorText.textContent).toBe('This field is required.');
  });
});
