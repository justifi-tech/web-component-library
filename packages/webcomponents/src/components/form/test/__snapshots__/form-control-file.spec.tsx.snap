// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form-control-file renders with default props 1`] = `
<form-control-file exportparts="label,input,input-invalid" label="Select a file" name="user ID">
  <div>
    <label class="form-label" htmlfor="user ID" part="label">
      Select a file
    </label>
    <input class="form-control" name="user ID" part="input " type="file">
    <form-control-help-text></form-control-help-text>
    <form-control-error-text></form-control-error-text>
  </div>
</form-control-file>
`;
